src/main/java/com/doctusoft/java/Retry.java

```diff
@@ -0,0 +1,132 @@
+     * @throws RuntimeException upon permanent failure (after retry attempts are exhausted)
+     */
+    public static <T> T callWithExponentialBackOff(int maxRetries, Callable<T> task) {
+        return callWithDelays(task, ExponentialDelays
+            .longsFrom(1000L)
+            .limitMaxAttempts(maxRetries + 1));
+    }
+    
+    /**
+     * Executes a {@link Runnable} task according to the provided exponential delays (in milliseconds).
+     *
+     * @param runnable the runnable task
+     * @param delays   builder for a delay values stream (in milliseconds)
+     * @throws RuntimeException upon permanent failure (after retry attempts are exhausted)
+     */
+    public static void runWithDelays(Runnable runnable, ExponentialDelays.OfLong delays) {
+        callWithDelays(Executors.callable(runnable), delays);
+    }
+    
+    /**
+     * Executes a {@link Runnable} task according to the default exponential back-off mechanism starting with a default
+     * 1s delay after the first attempt fails with the provided maximum number of retries.
+     *
+     * @param maxRetries the maximum number of retries
+     * @param runnable   the runnable task
+     * @throws RuntimeException upon permanent failure (after retry attempts are exhausted)
+     */
+    public static void runWithExponentialBackOff(int maxRetries, Runnable runnable) {
+        callWithExponentialBackOff(maxRetries, Executors.callable(runnable));
+    }
+    
+    private final Callable<T> task;
+    
+    private final Spliterator.OfLong delays;
+    
+    private int retryCount = 0;
+    
+    public Retry(Callable<T> task, LongStream delays) {
+        this.task = requireNonNull(task, "task");
+        this.delays = requireNonNull(delays, "delays").spliterator();
+    }
+    
+    public void run() {
+        call();
+    }
+    
+    public T call() throws RuntimeException {
+        while (true) {
+            try {
+                ++retryCount;
+                return task.call();
+            } catch (CancellationException e) {
+                log.log(Level.SEVERE, "Task canceled", e);
+                throw e;
+            } catch (Exception e) {
+                log.log(Level.WARNING, e, () -> retryCount + ". attempt failed");
+                if (!delays.tryAdvance((LongConsumer) this::delay)) {
+                    log.log(Level.SEVERE, "Failed permanently: no more retries");
+                    if (e instanceof RuntimeException) {
+                        throw (RuntimeException) e;
+                    }
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+    }
+    
+    protected void delay(long timeToWait) {
+        log.log(Level.INFO, () -> "Waiting " + printMillis(timeToWait) + " before retrying...");
+        try {
+            Thread.sleep(timeToWait);
+        } catch (InterruptedException ie) {
+            log.log(Level.FINE, "Interrupted", ie);
```

Comment:
Purpose was to log the interrupt event not to prevent the interrupt. Tried to make changes in order to avoid the latter.